// @oku-ui/primitives/dist/namespaced/index.cjs
'use strict';

const primitives = require('@oku-ui/primitives');

const Accordion = {
  Content: primitives.AccordionContent,
  Header: primitives.AccordionHeader,
  Item: primitives.AccordionItem,
  Root: primitives.AccordionRoot,
  Trigger: primitives.AccordionTrigger
};
const AlertDialog = {
  Root: primitives.AlertDialogRoot,
  Trigger: primitives.AlertDialogTrigger,
  Portal: primitives.AlertDialogPortal,
  Content: primitives.AlertDialogContent,
  Overlay: primitives.AlertDialogOverlay,
  Cancel: primitives.AlertDialogCancel,
  Title: primitives.AlertDialogTitle,
  Description: primitives.AlertDialogDescription,
  Action: primitives.AlertDialogAction
};
const Avatar = {
  Root: primitives.AvatarRoot,
  Image: primitives.AvatarImage,
  Fallback: primitives.AvatarFallback
};
const Checkbox = {
  Root: primitives.CheckboxRoot,
  Indicator: primitives.CheckboxIndicator
};
const Collapsible = {
  Root: primitives.CollapsibleRoot,
  Trigger: primitives.CollapsibleTrigger,
  Content: primitives.CollapsibleContent
};
const ContextMenu = {
  Root: primitives.ContextMenuRoot,
  Trigger: primitives.ContextMenuTrigger,
  Portal: primitives.ContextMenuPortal,
  Content: primitives.ContextMenuContent,
  Arrow: primitives.ContextMenuArrow,
  Item: primitives.ContextMenuItem,
  Group: primitives.ContextMenuGroup,
  Separator: primitives.ContextMenuSeparator,
  CheckboxItem: primitives.ContextMenuCheckboxItem,
  ItemIndicator: primitives.ContextMenuItemIndicator,
  Label: primitives.ContextMenuLabel,
  RadioGroup: primitives.ContextMenuRadioGroup,
  RadioItem: primitives.ContextMenuRadioItem,
  Sub: primitives.ContextMenuSub,
  SubContent: primitives.ContextMenuSubContent,
  SubTrigger: primitives.ContextMenuSubTrigger
};
const Dialog = {
  Root: primitives.DialogRoot,
  Trigger: primitives.DialogTrigger,
  Portal: primitives.DialogPortal,
  Content: primitives.DialogContent,
  Overlay: primitives.DialogOverlay,
  Close: primitives.DialogClose,
  Title: primitives.DialogTitle,
  Description: primitives.DialogDescription
};
const DropdownMenu = {
  Root: primitives.DropdownMenuRoot,
  Trigger: primitives.DropdownMenuTrigger,
  Portal: primitives.DropdownMenuPortal,
  Content: primitives.DropdownMenuContent,
  Arrow: primitives.DropdownMenuArrow,
  Item: primitives.DropdownMenuItem,
  Group: primitives.DropdownMenuGroup,
  Separator: primitives.DropdownMenuSeparator,
  CheckboxItem: primitives.DropdownMenuCheckboxItem,
  ItemIndicator: primitives.DropdownMenuItemIndicator,
  Label: primitives.DropdownMenuLabel,
  RadioGroup: primitives.DropdownMenuRadioGroup,
  RadioItem: primitives.DropdownMenuRadioItem,
  Sub: primitives.DropdownMenuSub,
  SubContent: primitives.DropdownMenuSubContent,
  SubTrigger: primitives.DropdownMenuSubTrigger
};
const HoverCard = {
  Root: primitives.HoverCardRoot,
  Trigger: primitives.HoverCardTrigger,
  Portal: primitives.HoverCardPortal,
  Content: primitives.HoverCardContent,
  Arrow: primitives.HoverCardArrow
};
const Menubar = {
  Root: primitives.MenubarRoot,
  Trigger: primitives.MenubarTrigger,
  Portal: primitives.MenubarPortal,
  Content: primitives.MenubarContent,
  Arrow: primitives.MenubarArrow,
  Item: primitives.MenubarItem,
  Group: primitives.MenubarGroup,
  Separator: primitives.MenubarSeparator,
  CheckboxItem: primitives.MenubarCheckboxItem,
  ItemIndicator: primitives.MenubarItemIndicator,
  Label: primitives.MenubarLabel,
  RadioGroup: primitives.MenubarRadioGroup,
  RadioItem: primitives.MenubarRadioItem,
  Sub: primitives.MenubarSub,
  SubContent: primitives.MenubarSubContent,
  SubTrigger: primitives.MenubarSubTrigger,
  Menu: primitives.MenubarMenu
};
const Popover = {
  Root: primitives.PopoverRoot,
  Trigger: primitives.PopoverTrigger,
  Portal: primitives.PopoverPortal,
  Content: primitives.PopoverContent,
  Arrow: primitives.PopoverArrow,
  Close: primitives.PopoverClose,
  Anchor: primitives.PopoverAnchor
};
const Progress = {
  Root: primitives.ProgressRoot,
  Indicator: primitives.ProgressIndicator
};
const RadioGroup = {
  Root: primitives.RadioGroupRoot,
  Item: primitives.RadioGroupItem,
  Indicator: primitives.RadioGroupIndicator
};
const ScrollArea = {
  Root: primitives.ScrollAreaRoot,
  Viewport: primitives.ScrollAreaViewport,
  Scrollbar: primitives.ScrollAreaScrollbar,
  Thumb: primitives.ScrollAreaThumb,
  Corner: primitives.ScrollAreaCorner
};
const Slider = {
  Root: primitives.SliderRoot,
  Thumb: primitives.SliderThumb,
  Track: primitives.SliderTrack,
  Range: primitives.SliderRange
};
const Switch = {
  Root: primitives.SwitchRoot,
  Thumb: primitives.SwitchThumb
};
const Tabs = {
  Root: primitives.TabsRoot,
  List: primitives.TabsList,
  Content: primitives.TabsContent,
  Trigger: primitives.TabsTrigger
};
const Toast = {
  Root: primitives.ToastRoot,
  Action: primitives.ToastAction,
  Close: primitives.ToastClose,
  Viewport: primitives.ToastViewport,
  Title: primitives.ToastTitle,
  Description: primitives.ToastDescription
};
const ToggleGroup = {
  Root: primitives.ToggleGroupRoot,
  Item: primitives.ToggleGroupItem
};
const Toolbar = {
  Root: primitives.ToolbarRoot,
  Button: primitives.ToolbarButton,
  Link: primitives.ToolbarLink,
  ToggleGroup: primitives.ToolbarToggleGroup,
  ToggleItem: primitives.ToolbarToggleItem,
  Separator: primitives.ToolbarSeparator
};
const Tooltip = {
  Root: primitives.TooltipRoot,
  Trigger: primitives.TooltipTrigger,
  Content: primitives.TooltipContent,
  Arrow: primitives.TooltipArrow,
  Portal: primitives.TooltipPortal,
  Provider: primitives.TooltipProvider
};

exports.AspectRatio = primitives.AspectRatio;
exports.Label = primitives.Label;
exports.Separator = primitives.Separator;
exports.Toggle = primitives.Toggle;
exports.Accordion = Accordion;
exports.AlertDialog = AlertDialog;
exports.Avatar = Avatar;
exports.Checkbox = Checkbox;
exports.Collapsible = Collapsible;
exports.ContextMenu = ContextMenu;
exports.Dialog = Dialog;
exports.DropdownMenu = DropdownMenu;
exports.HoverCard = HoverCard;
exports.Menubar = Menubar;
exports.Popover = Popover;
exports.Progress = Progress;
exports.RadioGroup = RadioGroup;
exports.ScrollArea = ScrollArea;
exports.Slider = Slider;
exports.Switch = Switch;
exports.Tabs = Tabs;
exports.Toast = Toast;
exports.ToggleGroup = ToggleGroup;
exports.Toolbar = Toolbar;
exports.Tooltip = Tooltip;
