// @oku-ui/primitives/dist/aspect-ratio.mjs
import { defineComponent, openBlock, createElementBlock, normalizeProps, guardReactiveProps, unref, createVNode, withCtx, renderSlot, mergeDefaults } from 'vue';
import { m as mergePrimitiveAttrs, c as convertPropsToHookProps, n as normalizeAttrs } from './shared.mjs';
import { P as Primitive } from './primitive.mjs';

const DEFAULT_ASPECT_RATIO_PROPS = {
  ratio: 1
};
function useAspectRatio(props = {}) {
  const { ratio = () => 1 } = props;
  return {
    wrapperAttrs() {
      const attrs = {
        "style": {
          // ensures inner element is contained
          position: "relative",
          // ensures padding bottom trick maths works
          width: "100%",
          paddingBottom: `${100 / ratio()}%`
        },
        "data-radix-aspect-ratio-wrapper": ""
      };
      return attrs;
    },
    attrs(extraAttrs) {
      const attrs = {
        style: {
          // ensures children expand in ratio
          position: "absolute",
          top: 0,
          right: 0,
          bottom: 0,
          left: 0
        }
      };
      if (extraAttrs && extraAttrs.length > 0) {
        mergePrimitiveAttrs(attrs, extraAttrs);
      }
      return attrs;
    }
  };
}

const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "AspectRatio",
    inheritAttrs: false
  },
  __name: "AspectRatio",
  props: /* @__PURE__ */ mergeDefaults({
    ratio: {}
  }, DEFAULT_ASPECT_RATIO_PROPS),
  setup(__props) {
    const props = __props;
    const aspectRatio = useAspectRatio(convertPropsToHookProps(props, ["ratio"]));
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", normalizeProps(guardReactiveProps(unref(aspectRatio).wrapperAttrs())), [
        createVNode(unref(Primitive), normalizeProps(guardReactiveProps(unref(normalizeAttrs)(unref(aspectRatio).attrs([_ctx.$attrs])))), {
          default: withCtx(() => [
            renderSlot(_ctx.$slots, "default")
          ]),
          _: 3
        }, 16)
      ], 16);
    };
  }
});

export { DEFAULT_ASPECT_RATIO_PROPS as D, _sfc_main as _ };
